lib/web_ui/lib/src/engine/html/shaders/image_shader.dart

```diff
@@ -2,20 +2,269 @@
+
+    int widthInPixels = (shaderBounds!.width * dpr).ceil();
+    int heightInPixels = (shaderBounds.height * dpr).ceil();
+
+    assert(widthInPixels > 0 && heightInPixels > 0);
+
+    /// Render tiles into a bitmap and create a canvas pattern.
+    final bool isWebGl2 = webGLVersion == WebGLVersion.webgl2;
+
+    final String vertexShader = VertexShaders.writeTextureVertexShader();
+    final String fragmentShader = FragmentShaders.writeTextureFragmentShader(
+        isWebGl2, tileModeX, tileModeY);
+
+    /// Render gradient into a bitmap and create a canvas pattern.
+    OffScreenCanvas offScreenCanvas =
+        OffScreenCanvas(widthInPixels, heightInPixels);
+    GlContext gl = GlContext(offScreenCanvas);
+    gl.setViewportSize(widthInPixels, heightInPixels);
+
+    GlProgram glProgram = gl.cacheProgram(vertexShader, fragmentShader);
+    gl.useProgram(glProgram);
+
+    final Float32List vertices = Float32List(12);
+    ui.Rect vRect = shaderBounds.translate(-shaderBounds.left, -shaderBounds.top);
+    vertices[0] = vRect.left;
+    vertices[1] = vRect.top;
+    vertices[2] = vRect.right;
+    vertices[3] = vRect.top;
+    vertices[4] = vRect.right;
+    vertices[5] = vRect.bottom;
+    vertices[6] = vRect.right;
+    vertices[7] = vRect.bottom;
+    vertices[8] = vRect.left;
+    vertices[9] = vRect.bottom;
+    vertices[10] = vRect.left;
+    vertices[11] = vRect.top;
+
+    Object? positionAttributeLocation =
+        gl.getAttributeLocation(glProgram.program, 'position');
+
+    setupVertexTransforms(gl, glProgram, 0, 0,
+        widthInPixels.toDouble(), heightInPixels.toDouble(), transform);
+
+    requiresTileOffset = shaderBounds.left !=0 || shaderBounds.top != 0;
+
+    /// To map from vertex position to texture coordinate in 0..1 range,
+    /// we setup scalar to be used in vertex shader.
+    setupTextureTransform(
+        gl,
+        glProgram,
+        shaderBounds.left,
+        shaderBounds.top,
+        1.0 / image.width.toDouble(),
+        1.0 / image.height.toDouble());
+
+    /// Setup geometry.
+    ///
+    /// Create buffer for vertex coordinates.
+    Object positionsBuffer = gl.createBuffer()!;
+    assert(positionsBuffer != null); // ignore: unnecessary_null_comparison
+
+    Object? vao;
+    if (isWebGl2) {
+      /// Create a vertex array object.
+      vao = gl.createVertexArray();
+      /// Set vertex array object as active one.
+      gl.bindVertexArray(vao!);
+    }
+
+    /// Turn on position attribute.
+    gl.enableVertexAttribArray(positionAttributeLocation);
+    /// Bind buffer as position buffer and transfer data.
+    gl.bindArrayBuffer(positionsBuffer);
+    bufferVertexData(gl, vertices, ui.window.devicePixelRatio);
+
+    /// Setup data format for attribute.
+    js_util.callMethod(gl.glContext, 'vertexAttribPointer', <dynamic>[
+      positionAttributeLocation,
+      2,
+      gl.kFloat,
+      false,
+      0,
+      0,
+    ]);
+
+    final int vertexCount = vertices.length ~/ 2;
```

Comment:
Does `bufferVertexData` change the size of `vertices`? If not, isn't this always `6`?

Reply:
Done. converted to const.